File: Sources/NBT/BedrockNBTCodec.swift
import Foundation

enum BedrockNBTCodec {
    static func decode(_ data: Data, encoding: NBTEncoding = .littleEndian, maximumDepth: Int = 256) throws -> NBTDocument {
        var cursor = BinaryCursor(data: data)
        let document = try decodeDocument(cursor: &cursor, encoding: encoding, maximumDepth: maximumDepth)
        return document
    }

    static func decodeDocument(cursor: inout BinaryCursor, encoding: NBTEncoding = .littleEndian, maximumDepth: Int = 256) throws -> NBTDocument {
        let rawType = try cursor.readByte()
        guard let type = NBTTagType(rawValue: rawType), type != .end else {
            throw BlocktopographError.malformedData("NBT 根标签类型无效：\(rawType)")
        }
        let name = try readString(cursor: &cursor, encoding: encoding)
        let value = try readPayload(type: type, cursor: &cursor, encoding: encoding, depth: 0, maximumDepth: maximumDepth)
        return NBTDocument(rootName: name, root: value)
    }

    static func encode(_ document: NBTDocument, encoding: NBTEncoding = .littleEndian) throws -> Data {
        var writer = BinaryWriter()
        writer.writeByte(document.root.type.rawValue)
        try writeString(document.rootName, writer: &writer, encoding: encoding)
        try writePayload(document.root, writer: &writer, encoding: encoding)
        return writer.data
    }

    private static func readPayload(type: NBTTagType, cursor: inout BinaryCursor, encoding: NBTEncoding, depth: Int, maximumDepth: Int) throws -> NBTValue {
        guard depth <= maximumDepth else { throw BlocktopographError.malformedData("NBT 嵌套超过 \(maximumDepth) 层") }
        switch type {
        case .end:
            throw BlocktopographError.malformedData("End 标签不能作为值")
        case .byte:
            return .byte(Int8(bitPattern: try cursor.readByte()))
        case .short:
            return .short(encoding == .bigEndian ? try cursor.readInt16BE() : try cursor.readInt16LE())
        case .int:
            switch encoding {
            case .bigEndian: return .int(try cursor.readInt32BE())
            case .littleEndian: return .int(try cursor.readInt32LE())
            case .littleEndianVarInt: return .int(try cursor.readSignedVarInt32())
            }
        case .long:
            switch encoding {
            case .bigEndian: return .long(try cursor.readInt64BE())
            case .littleEndian: return .long(try cursor.readInt64LE())
            case .littleEndianVarInt: return .long(try cursor.readSignedVarInt64())
            }
        case .float:
            return .float(encoding == .bigEndian ? try cursor.readFloatBE() : try cursor.readFloatLE())
        case .double:
            return .double(encoding == .bigEndian ? try cursor.readDoubleBE() : try cursor.readDoubleLE())
        case .byteArray:
            let count = try readLength(cursor: &cursor, encoding: encoding)
            return .byteArray(try cursor.readData(count: count))
        case .string:
            return .string(try readString(cursor: &cursor, encoding: encoding))
        case .list:
            let elementRaw = try cursor.readByte()
            guard let elementType = NBTTagType(rawValue: elementRaw) else {
                throw BlocktopographError.malformedData("NBT List 元素类型无效：\(elementRaw)")
            }
            let count = try readLength(cursor: &cursor, encoding: encoding)
            if count > 0 && elementType == .end {
                throw BlocktopographError.malformedData("非空 List 不能使用 End 元素类型")
            }
            var values = [NBTValue]()
            values.reserveCapacity(count)
            for _ in 0..<count {
                values.append(try readPayload(type: elementType, cursor: &cursor, encoding: encoding, depth: depth + 1, maximumDepth: maximumDepth))
            }
            return .list(elementType, values)
        case .compound:
            var tags = [NBTNamedTag]()
            while true {
                let raw = try cursor.readByte()
                guard let childType = NBTTagType(rawValue: raw) else {
                    throw BlocktopographError.malformedData("Compound 子标签类型无效：\(raw)")
                }
                if childType == .end { break }
                let name = try readString(cursor: &cursor, encoding: encoding)
                let value = try readPayload(type: childType, cursor: &cursor, encoding: encoding, depth: depth + 1, maximumDepth: maximumDepth)
                tags.append(NBTNamedTag(name: name, value: value))
            }
            return .compound(tags)
        case .intArray:
            let count = try readLength(cursor: &cursor, encoding: encoding)
            var values = [Int32]()
            values.reserveCapacity(count)
            for _ in 0..<count {
                switch encoding {
                case .bigEndian: values.append(try cursor.readInt32BE())
                case .littleEndian: values.append(try cursor.readInt32LE())
                case .littleEndianVarInt: values.append(try cursor.readSignedVarInt32())
                }
            }
            return .intArray(values)
        case .longArray:
            let count = try readLength(cursor: &cursor, encoding: encoding)
            var values = [Int64]()
            values.reserveCapacity(count)
            for _ in 0..<count {
                switch encoding {
                case .bigEndian: values.append(try cursor.readInt64BE())
                case .littleEndian: values.append(try cursor.readInt64LE())
                case .littleEndianVarInt: values.append(try cursor.readSignedVarInt64())
                }
            }
            return .longArray(values)
        }
    }

    private static func writePayload(_ value: NBTValue, writer: inout BinaryWriter, encoding: NBTEncoding) throws {
        switch value {
        case .byte(let number):
            writer.writeByte(UInt8(bitPattern: number))
        case .short(let number):
            if encoding == .bigEndian { writer.writeInt16BE(number) } else { writer.writeInt16LE(number) }
        case .int(let number):
            switch encoding {
            case .bigEndian: writer.writeInt32BE(number)
            case .littleEndian: writer.writeInt32LE(number)
            case .littleEndianVarInt: writer.writeSignedVarInt(number)
            }
        case .long(let number):
            switch encoding {
            case .bigEndian: writer.writeInt64BE(number)
            case .littleEndian: writer.writeInt64LE(number)
            case .littleEndianVarInt: writer.writeSignedVarLong(number)
            }
        case .float(let number):
            if encoding == .bigEndian { writer.writeFloatBE(number) } else { writer.writeFloatLE(number) }
        case .double(let number):
            if encoding == .bigEndian { writer.writeDoubleBE(number) } else { writer.writeDoubleLE(number) }
        case .byteArray(let bytes):
            try writeLength(bytes.count, writer: &writer, encoding: encoding)
            writer.writeData(bytes)
        case .string(let text):
            try writeString(text, writer: &writer, encoding: encoding)
        case .list(let elementType, let values):
            guard values.allSatisfy({ $0.type == elementType }) else {
                throw BlocktopographError.malformedData("NBT List 中存在不同类型元素")
            }
            writer.writeByte(elementType.rawValue)
            try writeLength(values.count, writer: &writer, encoding: encoding)
            for child in values { try writePayload(child, writer: &writer, encoding: encoding) }
        case .compound(let tags):
            for tag in tags {
                writer.writeByte(tag.value.type.rawValue)
                try writeString(tag.name, writer: &writer, encoding: encoding)
                try writePayload(tag.value, writer: &writer, encoding: encoding)
            }
            writer.writeByte(NBTTagType.end.rawValue)
        case .intArray(let values):
            try writeLength(values.count, writer: &writer, encoding: encoding)
            for number in values {
                switch encoding {
                case .bigEndian: writer.writeInt32BE(number)
                case .littleEndian: writer.writeInt32LE(number)
                case .littleEndianVarInt: writer.writeSignedVarInt(number)
                }
            }
        case .longArray(let values):
            try writeLength(values.count, writer: &writer, encoding: encoding)
            for number in values {
                switch encoding {
                case .bigEndian: writer.writeInt64BE(number)
                case .littleEndian: writer.writeInt64LE(number)
                case .littleEndianVarInt: writer.writeSignedVarLong(number)
                }
            }
        }
    }

    private static func readLength(cursor: inout BinaryCursor, encoding: NBTEncoding) throws -> Int {
        let raw: Int64
        switch encoding {
        case .bigEndian: raw = Int64(try cursor.readInt32BE())
        case .littleEndian: raw = Int64(try cursor.readInt32LE())
        case .littleEndianVarInt: raw = Int64(try cursor.readSignedVarInt32())
        }
        guard raw >= 0, raw <= Int64(Int.max), raw <= 100_000_000 else {
            throw BlocktopographError.malformedData("NBT 长度无效：\(raw)")
        }
        return Int(raw)
    }

    private static func writeLength(_ count: Int, writer: inout BinaryWriter, encoding: NBTEncoding) throws {
        guard count >= 0, count <= Int(Int32.max) else {
            throw BlocktopographError.malformedData("NBT 长度无法编码：\(count)")
        }
        switch encoding {
        case .bigEndian: writer.writeInt32BE(Int32(count))
        case .littleEndian: writer.writeInt32LE(Int32(count))
        case .littleEndianVarInt: writer.writeSignedVarInt(Int32(count))
        }
    }

    private static func readString(cursor: inout BinaryCursor, encoding: NBTEncoding) throws -> String {
        let length: Int
        switch encoding {
        case .bigEndian:
            length = Int(try cursor.readUInt16BE())
        case .littleEndian:
            length = Int(try cursor.readUInt16LE())
        case .littleEndianVarInt:
            let value = try cursor.readUnsignedVarInt(maxBytes: 5)
            guard value <= UInt64(Int.max), value <= 16_777_216 else {
                throw BlocktopographError.malformedData("NBT 字符串过长")
            }
            length = Int(value)
        }
        let data = try cursor.readData(count: length)
        guard let text = String(data: data, encoding: .utf8) else {
            throw BlocktopographError.malformedData("NBT 字符串不是 UTF-8")
        }
        return text
    }

    private static func writeString(_ value: String, writer: inout BinaryWriter, encoding: NBTEncoding) throws {
        guard let data = value.data(using: .utf8) else {
            throw BlocktopographError.malformedData("字符串无法编码为 UTF-8")
        }
        switch encoding {
        case .bigEndian:
            guard data.count <= Int(UInt16.max) else { throw BlocktopographError.malformedData("NBT 字符串超过 65535 字节") }
            writer.writeUInt16BE(UInt16(data.count))
        case .littleEndian:
            guard data.count <= Int(UInt16.max) else { throw BlocktopographError.malformedData("NBT 字符串超过 65535 字节") }
            writer.writeUInt16LE(UInt16(data.count))
        case .littleEndianVarInt:
            writer.writeUnsignedVarInt(UInt64(data.count))
        }
        writer.writeData(data)
    }
}

File: Windows/MCBEEditor.Core/Nbt/BedrockNbtCodec.cs
namespace MCBEEditor.Core.Nbt;

public static class BedrockNbtCodec
{
    public static NbtDocument Decode(ReadOnlySpan<byte> data, NbtEncoding encoding = NbtEncoding.LittleEndian, int maximumDepth = 256)
        => DecodeOne(data, out _, encoding, maximumDepth);

    /// <summary>
    /// Decodes one NBT document from the start of <paramref name="data"/> and
    /// reports exactly how many bytes were consumed. Bedrock SubChunk palettes
    /// store consecutive unnamed NBT compounds, so callers must be able to
    /// continue at the byte immediately following one document.
    /// </summary>
    public static NbtDocument DecodeOne(
        ReadOnlySpan<byte> data,
        out int bytesRead,
        NbtEncoding encoding = NbtEncoding.LittleEndian,
        int maximumDepth = 256)
    {
        var reader = new BinaryDataReader(data);
        var rawType = reader.ReadByte();
        if (!Enum.IsDefined(typeof(NbtTagType), rawType) || rawType == (byte)NbtTagType.End)
            throw new InvalidDataException($"NBT 根标签类型无效：{rawType}");

        var type = (NbtTagType)rawType;
        var name = ReadString(ref reader, encoding);
        var value = ReadPayload(type, ref reader, encoding, 0, maximumDepth);
        bytesRead = reader.Offset;
        return new NbtDocument(name, value);
    }

    public static byte[] Encode(NbtDocument document, NbtEncoding encoding = NbtEncoding.LittleEndian)
    {
        var writer = new BinaryDataWriter();
        writer.WriteByte((byte)document.Root.Type);
        WriteString(document.RootName, writer, encoding);
        WritePayload(document.Root, writer, encoding);
        return writer.ToArray();
    }

    private static NbtValue ReadPayload(NbtTagType type, ref BinaryDataReader reader, NbtEncoding encoding, int depth, int maximumDepth)
    {
        if (depth > maximumDepth)
            throw new InvalidDataException($"NBT 嵌套超过 {maximumDepth} 层");

        return type switch
        {
            NbtTagType.End => throw new InvalidDataException("End 标签不能作为值"),
            NbtTagType.Byte => new NbtByteValue(unchecked((sbyte)reader.ReadByte())),
            NbtTagType.Short => new NbtShortValue(encoding == NbtEncoding.BigEndian ? reader.ReadInt16BE() : reader.ReadInt16LE()),
            NbtTagType.Int => new NbtIntValue(ReadInt(ref reader, encoding)),
            NbtTagType.Long => new NbtLongValue(ReadLong(ref reader, encoding)),
            NbtTagType.Float => new NbtFloatValue(encoding == NbtEncoding.BigEndian ? reader.ReadFloatBE() : reader.ReadFloatLE()),
            NbtTagType.Double => new NbtDoubleValue(encoding == NbtEncoding.BigEndian ? reader.ReadDoubleBE() : reader.ReadDoubleLE()),
            NbtTagType.ByteArray => new NbtByteArrayValue(reader.ReadBytes(ReadLength(ref reader, encoding))),
            NbtTagType.String => new NbtStringValue(ReadString(ref reader, encoding)),
            NbtTagType.List => ReadList(ref reader, encoding, depth, maximumDepth),
            NbtTagType.Compound => ReadCompound(ref reader, encoding, depth, maximumDepth),
            NbtTagType.IntArray => ReadIntArray(ref reader, encoding),
            NbtTagType.LongArray => ReadLongArray(ref reader, encoding),
            _ => throw new InvalidDataException($"不支持的 NBT 类型：{type}")
        };
    }

    private static NbtListValue ReadList(ref BinaryDataReader reader, NbtEncoding encoding, int depth, int maximumDepth)
    {
        var rawElementType = reader.ReadByte();
        if (!Enum.IsDefined(typeof(NbtTagType), rawElementType))
            throw new InvalidDataException($"NBT List 元素类型无效：{rawElementType}");

        var elementType = (NbtTagType)rawElementType;
        var count = ReadLength(ref reader, encoding);
        if (elementType == NbtTagType.End && count != 0)
            throw new InvalidDataException("NBT List 的 End 元素类型只允许用于空列表");

        var values = new NbtValue[count];
        for (var index = 0; index < count; index++)
            values[index] = ReadPayload(elementType, ref reader, encoding, depth + 1, maximumDepth);
        return new NbtListValue(elementType, values);
    }

    private static NbtCompoundValue ReadCompound(ref BinaryDataReader reader, NbtEncoding encoding, int depth, int maximumDepth)
    {
        var tags = new List<NbtNamedTag>();
        while (true)
        {
            var rawType = reader.ReadByte();
            if (rawType == (byte)NbtTagType.End)
                break;
            if (!Enum.IsDefined(typeof(NbtTagType), rawType))
                throw new InvalidDataException($"NBT Compound 子标签类型无效：{rawType}");

            var childType = (NbtTagType)rawType;
            var name = ReadString(ref reader, encoding);
            var value = ReadPayload(childType, ref reader, encoding, depth + 1, maximumDepth);
            tags.Add(new NbtNamedTag(name, value));
        }
        return new NbtCompoundValue(tags);
    }

    private static NbtIntArrayValue ReadIntArray(ref BinaryDataReader reader, NbtEncoding encoding)
    {
        var count = ReadLength(ref reader, encoding);
        var values = new int[count];
        for (var i = 0; i < count; i++)
            values[i] = ReadInt(ref reader, encoding);
        return new NbtIntArrayValue(values);
    }

    private static NbtLongArrayValue ReadLongArray(ref BinaryDataReader reader, NbtEncoding encoding)
    {
        var count = ReadLength(ref reader, encoding);
        var values = new long[count];
        for (var i = 0; i < count; i++)
            values[i] = ReadLong(ref reader, encoding);
        return new NbtLongArrayValue(values);
    }

    private static int ReadInt(ref BinaryDataReader reader, NbtEncoding encoding) => encoding switch
    {
        NbtEncoding.BigEndian => reader.ReadInt32BE(),
        NbtEncoding.LittleEndian => reader.ReadInt32LE(),
        NbtEncoding.LittleEndianVarInt => reader.ReadSignedVarInt32(),
        _ => throw new ArgumentOutOfRangeException(nameof(encoding))
    };

    private static long ReadLong(ref BinaryDataReader reader, NbtEncoding encoding) => encoding switch
    {
        NbtEncoding.BigEndian => reader.ReadInt64BE(),
        NbtEncoding.LittleEndian => reader.ReadInt64LE(),
        NbtEncoding.LittleEndianVarInt => reader.ReadSignedVarInt64(),
        _ => throw new ArgumentOutOfRangeException(nameof(encoding))
    };

    private static int ReadLength(ref BinaryDataReader reader, NbtEncoding encoding)
    {
        long raw = encoding switch
        {
            NbtEncoding.BigEndian => reader.ReadInt32BE(),
            NbtEncoding.LittleEndian => reader.ReadInt32LE(),
            NbtEncoding.LittleEndianVarInt => reader.ReadSignedVarInt32(),
            _ => throw new ArgumentOutOfRangeException(nameof(encoding))
        };

        if (raw < 0 || raw > 100_000_000 || raw > int.MaxValue)
            throw new InvalidDataException($"NBT 长度无效：{raw}");
        return (int)raw;
    }

    private static string ReadString(ref BinaryDataReader reader, NbtEncoding encoding)
    {
        int length = encoding switch
        {
            NbtEncoding.BigEndian => reader.ReadUInt16BE(),
            NbtEncoding.LittleEndian => reader.ReadUInt16LE(),
            NbtEncoding.LittleEndianVarInt => ReadVarStringLength(ref reader),
            _ => throw new ArgumentOutOfRangeException(nameof(encoding))
        };

        return NbtRawStringCodec.Decode(reader.ReadBytes(length));
    }

    private static int ReadVarStringLength(ref BinaryDataReader reader)
    {
        var value = reader.ReadUnsignedVarInt(5);
        if (value > 16_777_216 || value > int.MaxValue)
            throw new InvalidDataException("NBT 字符串过长");
        return (int)value;
    }

    private static void WritePayload(NbtValue value, BinaryDataWriter writer, NbtEncoding encoding)
    {
        switch (value)
        {
            case NbtByteValue item:
                writer.WriteByte(unchecked((byte)item.Value));
                break;
            case NbtShortValue item:
                if (encoding == NbtEncoding.BigEndian) writer.WriteInt16BE(item.Value); else writer.WriteInt16LE(item.Value);
                break;
            case NbtIntValue item:
                WriteInt(item.Value, writer, encoding);
                break;
            case NbtLongValue item:
                WriteLong(item.Value, writer, encoding);
                break;
            case NbtFloatValue item:
                if (encoding == NbtEncoding.BigEndian) writer.WriteFloatBE(item.Value); else writer.WriteFloatLE(item.Value);
                break;
            case NbtDoubleValue item:
                if (encoding == NbtEncoding.BigEndian) writer.WriteDoubleBE(item.Value); else writer.WriteDoubleLE(item.Value);
                break;
            case NbtByteArrayValue item:
                WriteLength(item.Value.Length, writer, encoding);
                writer.WriteBytes(item.Value);
                break;
            case NbtStringValue item:
                WriteString(item.Value, writer, encoding);
                break;
            case NbtListValue item:
                writer.WriteByte((byte)item.ElementType);
                WriteLength(item.Values.Count, writer, encoding);
                foreach (var child in item.Values)
                {
                    if (child.Type != item.ElementType)
                        throw new InvalidDataException($"NBT List 声明类型 {item.ElementType}，但实际包含 {child.Type}");
                    WritePayload(child, writer, encoding);
                }
                break;
            case NbtCompoundValue item:
                foreach (var tag in item.Tags)
                {
                    writer.WriteByte((byte)tag.Value.Type);
                    WriteString(tag.Name, writer, encoding);
                    WritePayload(tag.Value, writer, encoding);
                }
                writer.WriteByte((byte)NbtTagType.End);
                break;
            case NbtIntArrayValue item:
                WriteLength(item.Values.Count, writer, encoding);
                foreach (var number in item.Values) WriteInt(number, writer, encoding);
                break;
            case NbtLongArrayValue item:
                WriteLength(item.Values.Count, writer, encoding);
                foreach (var number in item.Values) WriteLong(number, writer, encoding);
                break;
            default:
                throw new InvalidDataException($"不支持写入的 NBT 类型：{value.GetType().Name}");
        }
    }

    private static void WriteInt(int value, BinaryDataWriter writer, NbtEncoding encoding)
    {
        switch (encoding)
        {
            case NbtEncoding.BigEndian: writer.WriteInt32BE(value); break;
            case NbtEncoding.LittleEndian: writer.WriteInt32LE(value); break;
            case NbtEncoding.LittleEndianVarInt: writer.WriteSignedVarInt(value); break;
            default: throw new ArgumentOutOfRangeException(nameof(encoding));
        }
    }

    private static void WriteLong(long value, BinaryDataWriter writer, NbtEncoding encoding)
    {
        switch (encoding)
        {
            case NbtEncoding.BigEndian: writer.WriteInt64BE(value); break;
            case NbtEncoding.LittleEndian: writer.WriteInt64LE(value); break;
            case NbtEncoding.LittleEndianVarInt: writer.WriteSignedVarLong(value); break;
            default: throw new ArgumentOutOfRangeException(nameof(encoding));
        }
    }

    private static void WriteLength(int count, BinaryDataWriter writer, NbtEncoding encoding)
    {
        if (count < 0)
            throw new InvalidDataException($"NBT 长度无法编码：{count}");
        WriteInt(count, writer, encoding);
    }

    private static void WriteString(string value, BinaryDataWriter writer, NbtEncoding encoding)
    {
        var bytes = NbtRawStringCodec.Encode(value);
        switch (encoding)
        {
            case NbtEncoding.BigEndian:
                if (bytes.Length > ushort.MaxValue) throw new InvalidDataException("NBT 字符串超过 65535 字节");
                writer.WriteUInt16BE((ushort)bytes.Length);
                break;
            case NbtEncoding.LittleEndian:
                if (bytes.Length > ushort.MaxValue) throw new InvalidDataException("NBT 字符串超过 65535 字节");
                writer.WriteUInt16LE((ushort)bytes.Length);
                break;
            case NbtEncoding.LittleEndianVarInt:
                writer.WriteUnsignedVarInt((ulong)bytes.Length);
                break;
            default:
                throw new ArgumentOutOfRangeException(nameof(encoding));
        }
        writer.WriteBytes(bytes);
    }
}
